design/authn_azure/authn_azure_solution_design.md

```diff
@@ -0,0 +1,242 @@
+| Service principle                      | object representation of VM instance in AD. This object holds information defining its permissions on resources                                                                                 |
+| Client ID                              | unique identifier generated by Azure AD tied to application and service principal during initial provisioning. This client ID is used to identify the object that is requesting an access token |
+| System-assigned identity               | an identity for an instance in AD that is trusted. The identity is tied to the resource and abides to its lifecycle. In other words, if the resource is delete so to will the identity.         |
+| User-assigned identity                 | an identity for an instance in AD that is trusted. This identity can be assigned to 1+ Azure services instances and is independent of instance lifecycle                                        |
+| Azure Instance Metadata Service (IMDS) | a REST endpoint accessible to all VMs created via the Azure Resource Manager. Resources in Azure request access tokens from the Azure Instance Metadata service endpoint                        |
+| JWT                                    | JSON-based access token that asserts claims (logged in, etc)         
+
+## Useful links
+
+- [Azure Spike Ramp-up training exercise](https://ca-il-confluence.il.cyber-ark.com/display/rndp/Azure+Spike+Ramp-up#AzureSpikeRamp-up-/)
+
+- [Delivery Plan - Conjur Azure Authenticator](https://ca-il-confluence.il.cyber-ark.com/display/rndp/Delivery+Plan+-+Conjur+Azure+Authenticator)
+
+## Issue description
+
+Many of our customers are using Azure as their main cloud provider. Therefore we want to enable Azure resources to authenticate according to Azure 
+properties and get back a Conjur access token in order to retrieve secrets from our providers. At current, customers can authenticate their Azure 
+VMs with Conjur by creating a Conjur host identity for their Azure resources. The Azure resource would then receive an API key as part of the 
+creation of a Conjur identity which they must use to connect to targets. This is not optimal nor convenient because we are not allowing the 
+resources to authenticate based on their Azure identity and demand that the resource hold an extra secret, the Conjur-given API key.
+
+## Solution
+
+We will add a new authenticator to Conjur, which users can authenticate with from Azure VMs. As mentioned in the feature doc, before any authentication 
+request is sent to Conjur, the admin will load the authenticator policy: 
+
+```yaml
+# policy id needs to match the convention `conjur/authn-azure/<service ID>`
+- !policy
+  id: conjur/authn-azure/prod
+  body:
+  - !webservice
+    - !variable
+        id: provider-uri # https://sts.windows.net/TENANT_ID
```

Comment:
Is there any way this could be an annotation? Or is this limited because of OIDC implementation?
If it were an annotation it would help with UX.

Reply:
If I am understanding correctly, you are proposing adding the `provider-uri` as an annotation in the webservice instead of housing it as variable in Conjur. I appreciate you challenging this!  

In terms of security, I am not sure how sensitive this variable is and this is something we should definitely check @shaharglazner 

In terms of UX, this is how we configure the OIDC authenticator. I imagine that our users use multiple authenticators across their environments and so I am wary to suddenly change this configuration _just_ for Azure and not also have this change cascaded to OIDC.

@InbalZilberman @Tovli @orenbm opinions on this one?

Reply:
@AndrewCopeland this may have also performance aspects.
From UX I agree with Andrew it would be easier for the user.

Reply:
The only information that could be sensitive is the tenet ID, the azure STS URI looks like:
`https://sts.windows.net/<tenant ID>/`

It appears the tenant ID is similar to your account number in AWS.

Reply:
@shaharglazner @InbalZilberman if this data isn't sensitive then I'd be happy to change it to an annotation of the webservice

Reply:
Don't we want to give similar experiences across our authenticators?

Reply:
@orenbm  - no performance aspects?

Reply:
I didn't test this but I don't think there should be any performance issue.

Reply:
I think that if we decide that this is a better experience than the one defined in `authn-oidc` we should go for it. In the future we can add this ability to oidc as well (with backwards-compatibility with variables).

Reply:
@InbalZilberman wdyt?

Reply:
also - if `https://sts.windows.net/` is constant, we can let the user input only the tenant id. So we can have 
```
- !variable
        id: tenant-id 
```
And have `https://sts.windows.net/` hard-coded in the authenticator. @InbalZilberman wdyt?

Reply:
I'd like to summarize the options we have:
1. provider uri as variable:
```
- !policy
  id: conjur/authn-azure/prod
  body:
  - !webservice

  - !variable
    id: provider-uri
```
2. Provider URI as annotation:
```
```
- !policy
  id: conjur/authn-azure/prod
  body:
  - !webservice
    annotations:
      provider-uri: https://sts.windows.net/TENANT_ID
```
Some notes:
- UX: In option 1 the user also needs to load a value to the `provider-uri` variable so the UX is better in option 2. 
- Security: Option 1 holds the data as a secret but I don't see a security issue here as this data is not sensitive (but maybe @shaharglazner can correct me). So i think option 1 is fine with its security.
- Performance: It is mainly choosing wether to retrieve a secret from the DB or to retrieve a webservice's annotation (the webservice is already retrieved). I don't see a difference here.
- Implementation: Both options are generally developed. We may need to tweak some of the code but it shouldn't be too much.

In conclusion, i think we should go with option 2. 
@InbalZilberman @sigalsax @shaharglazner i'd like to get your comments on this.

Reply:
@orenbm @sigalsax @InbalZilberman @andrew
As far as I can find, tenant id is not sensitive - 

- https://stackoverflow.com/questions/45661109/are-azure-subscription-id-aad-tenant-id-and-aad-app-client-id-considered-secret

- https://stackoverflow.com/questions/56016004/are-the-azure-client-id-tenant-and-key-vault-uri-considered-secrets

Reply:
As a heads up, you cannot add annotations to a resources that does not have an ID 😢 
```
- !webservice
  annotations:
    provider-uri: https://sts.io
```
When the policy is loaded we get the following error:
```
{"error":{"code":"validation_failed","message":"policy_text webservice has a blank id","details":[{"code":"validation_failed","target":"policy_text","message":"webservice has a blank id"}]}}
```

Reply:
You don't have to give the webservice an ID if you have a policy like this
```
- !policy 
  id: policytest
  body:
    - !webservice
      annotations:
        provider-uri: https://sts.io
```
B/c the webservice will take the name of the policy ID
Ex: `conjur list` 
>"cucumber:webservice:policytest"

Reply:
@sigalsax so we will go with the annotation?

Reply:
We decided against annotation for maintenance concerns and will readdress this at a later date